pkg/connector/projects.go

```diff
@@ -0,0 +1,118 @@
+package connector
+
+import (
+	"context"
+	"errors"
+	"fmt"
+
+	"cloud.google.com/go/bigquery"
+	resourcemanager "cloud.google.com/go/resourcemanager/apiv3"
+	"cloud.google.com/go/resourcemanager/apiv3/resourcemanagerpb"
+	v2 "github.com/conductorone/baton-sdk/pb/c1/connector/v2"
+	"github.com/conductorone/baton-sdk/pkg/annotations"
+	"github.com/conductorone/baton-sdk/pkg/pagination"
+	rs "github.com/conductorone/baton-sdk/pkg/types/resource"
+	"google.golang.org/api/iterator"
+)
+
+type projectBuilder struct {
+	resourceType   *v2.ResourceType
+	projectsClient *resourcemanager.ProjectsClient
+	bigQueryClient *bigquery.Client
+}
+
+func (p *projectBuilder) ResourceType(ctx context.Context) *v2.ResourceType {
+	return projectResourceType
+}
+
+func projectResource(projects *resourcemanagerpb.Project) (*v2.Resource, error) {
+	var opts []rs.ResourceOption
+	profile := map[string]interface{}{
+		"id":          projects.ProjectId,
+		"name":        projects.Name,
+		"displayName": projects.DisplayName,
+	}
+
+	projectTraitOptions := []rs.AppTraitOption{
+		rs.WithAppProfile(profile),
+	}
+
+	opts = append(opts, rs.WithAppTrait(projectTraitOptions...))
+	resource, err := rs.NewResource(
+		projects.DisplayName,
+		projectResourceType,
+		projects.ProjectId,
+		opts...,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	return resource, nil
+}
+
+func (p *projectBuilder) List(ctx context.Context, parentResourceID *v2.ResourceId, pToken *pagination.Token) ([]*v2.Resource, string, annotations.Annotations, error) {
+	var (
+		resources []*v2.Resource
+		bag       = &pagination.Bag{}
+	)
+	err := bag.Unmarshal(pToken.Token)
+	if err != nil {
+		return nil, "", nil, err
+	}
+
+	if bag.Current() == nil {
+		bag.Push(pagination.PageState{
+			ResourceTypeID: projectResourceType.Id,
+		})
+	}
+
+	it := p.projectsClient.SearchProjects(ctx,
+		&resourcemanagerpb.SearchProjectsRequest{
+			Query:     "",
+			PageToken: bag.PageToken(),
+		},
+	)
+
+	for {
+		projects, err := it.Next()
+		if errors.Is(err, iterator.Done) {
+			break
+		}
+
+		resource, err := projectResource(projects)
+		if err != nil {
+			return nil, "", nil, wrapError(err, "Unable to create project resource")
+		}
+
+		resources = append(resources, resource)
+	}
+
+	err = bag.Next(it.PageInfo().Token)
+	if err != nil {
+		return nil, "", nil, fmt.Errorf("failed to fetch bag.Next: %w", err)
+	}
+
+	pageToken, err := bag.Marshal()
+	if err != nil {
+		return nil, "", nil, err
+	}
+
+	return resources, pageToken, nil, nil
+}
+
+func (p *projectBuilder) Entitlements(_ context.Context, resource *v2.Resource, _ *pagination.Token) ([]*v2.Entitlement, string, annotations.Annotations, error) {
+	return nil, "", nil, nil
```

Comment:
We are going to sync these as visible resources and resource types, but we won't be able to establish who has access to these resources without entitlements, and that's left as a todo. There was a configuration or permission issue that blocked this.

Reply:
I could include projects entitlements, but I would need investigate how those grants works in the case of projects.